ssl/quic/quic_port.c

```diff
@@ -708,23 +864,40 @@ static void port_send_retry(QUIC_PORT *port,
      */
     ok = ossl_quic_lcidm_get_unused_cid(port->lcidm, &hdr.src_conn_id);
     if (ok == 0)
-        return;
+        goto err;
+
+    /* Generate retry validation token */
+    if (!generate_retry_token(peer, client_hdr->dst_conn_id,
+                              client_hdr->src_conn_id, &token)
+        || !marshal_validation_token(&token, &token_buf, &token_buf_len))
+        goto err;
+
+    /* Make space for integrity tag */
+    token_buf_len += QUIC_RETRY_INTEGRITY_TAG_LEN;
+    tmp_buf = OPENSSL_realloc(token_buf, token_buf_len);
+    if (tmp_buf != NULL) {
+        token_buf = tmp_buf;
+        tmp_buf = NULL;
+    } else {
+        goto err;
+    }
 
     hdr.dst_conn_id = client_hdr->src_conn_id;
     hdr.type = QUIC_PKT_TYPE_RETRY;
     hdr.fixed = 1;
     hdr.version = 1;
-    hdr.len = token_len;
-    hdr.data = (unsigned char *)&token;
+    hdr.len = token_buf_len;
+    hdr.data = token_buf;
```

Comment:
I think the rscid is causing your ubsan failure in CI here.  hdr is declared on the stack, uninitialized, and so hdr.src_conn_id is whatever garbage was on the stack when the call was made.   ossl_quic_lcidm_get_unused_cid initalizes an unused cid, but only fills out the bytes in the id array which we use (currently 8).  The remaining 12 bytes are left as garbage values, and so the ub sanitizer is complaining that the retry tag integrity calculation is attempting to read those 12 bytes that have never been written.  

If you only marshal the cid.id_len bytes into the token, you can avoid this.  The alternative is to make sure every use of an QUIC_CONN_ID is initialized to a known value over the size of the entire struct.

Reply:
Ah ok, I'm hoping my recent commit fixes these issues (since I'm only copying the 8 bytes). Otherwise, I'm thinking I could do `hdr = { 0 }`

Reply:
I don't think you can just initalize hdr here on validation, as the parsing populates token.rscid from the token value, which only contains id_len bytes.  You would also need to init the token.rscid struct to all zeros to get a valid comparison.